src/FluentValidation/Internal/MessageFormatter.cs

```diff
@@ -93,21 +92,26 @@ public MessageFormatter AppendPropertyValue(object value)
 		/// </summary>
 		public Dictionary<string, object> PlaceholderValues => _placeholderValues;
 
-		protected virtual string ReplacePlaceholderWithValue(string template, string key, object value) {
-			string placeholder =  GetPlaceholder(key);
-			return template.Replace(placeholder, value?.ToString());
-		}
+		// TODO: Add alignment component
+		private Regex _keyRegex = new Regex("{([^{}:]+)(?::([^{}]+))?}"); // Matches any placeholder
 
-		protected string GetPlaceholder(string key) {
-			// Performance: String concat causes much overhead when not needed. Concatting constants results in constants being compiled.
-			switch (key) {
-				case PropertyName:
-					return "{" + PropertyName + "}";
-				case PropertyValue:
-					return "{" + PropertyValue + "}";
-				default:
-					return "{" + key + "}";
-			}
+		protected virtual string ReplacePlaceholdersWithValues(string template, IDictionary<string, object> values)
+		{
+			return _keyRegex.Replace(template, m =>
+			{
+				var key = m.Groups[1].Value;
+
+				if (!values.ContainsKey(key))
+					return m.Value; // No placeholder / value
+
+				var format = m.Groups[2].Success // Format specified?
+					? $"{{0:{m.Groups[2].Value}}}"
+					: null;
+
+				return format == null
+					? values[key]?.ToString()

```

Comment:
How does this handle the situation where a placeholder exists in the template, but no value has been specified? Guessing this will throw a key not found exception. The old behaviour would leave the placeholder in the string.

Could we have a test to make sure that works correctly?

Reply:
Doesn't this test do that?

		public void Should_ignore_unknown_parameters() {
			var now = DateTime.Now;
			string result = formatter
				.AppendArgument("foo", now)
				.BuildMessage("{foo:g} {unknown} {unknown:format}");

			result.ShouldEqual($"{(now.ToString("g"))} {{unknown}} {{unknown:format}}");
		}

Reply:
You’re right, I missed that one.